src/each/each.py

```diff
@@ -24,17 +27,164 @@
+
+
+@attr.s()
+class FileWorkItem(WorkItem):
+    """A file to be processed by ``Each``."""
+
+    """A name for this work item that can be used as a filename."""
+    name = attr.ib()
+
+    """The location of the file on disk."""
+    path = attr.ib()
+
+    def exists(self):
+        """A file work item exists only if the file exists."""
+        return os.path.exists(self.path)
+
+    def as_input_file(self):
+        return os.open(self.path, os.O_RDONLY)
+
+    def as_argument(self):
+        return os.path.abspath(self.path)
+
+    def write_in_file(self, path):
+        """The ``in`` file is a symlink to the original file."""
+        os.symlink(os.path.abspath(self.path), path)
+
+
+@attr.s()
+class LineWorkItem(WorkItem):
+    """A line to be processed by ``Each``."""
+
+    """A name for this work item that can be used as a filename."""
+    name = attr.ib()
+
+    """The line itself, including line ending, as text."""
+    line = attr.ib()
+
+    def exists(self):
+        """Whether or not this work item still exists.
+
+        A line always exists.
+        """
+        return True
+
+    def as_input_file(self):
+        r, w = os.pipe()
+        os.write(w, self.line.encode("utf-8"))
+        return r
+
+    def as_argument(self):
+        return self.line.rstrip("\n")
+
+    def write_in_file(self, path):
+        """The ``in`` file for a line work item is a file with just that line."""
+        with open(path, "w") as in_file:
+            in_file.write(self.line)
+
+
+def work_items_from_path(path):
+    """Load work items from a user-supplied path.
+
+    If 'path' is a directory, our work items are files. If it's a # file, our
+    work items are lines.
+    """
+    try:
+        return work_items_from_directory(path)
+    except NotADirectoryError:
+        with open(path, "r") as stream:
+            return work_items_from_file(stream)
+
+
+def work_items_from_directory(path):
+    """Get a list of work items from a directory.
+
+    Each file in the directory is a work item.
+    """
+    return (FileWorkItem(name=s, path=os.path.join(path, s)) for s in os.listdir(path))
+
+
+"""What we consider a 'simple name'.
+
+These are appended to the hashes of lines when generating output,
+as a convenience to the humans who must contend with it.
+"""
+simple_name_re = re.compile("^[A-Za-z0-9_-]+$")
+
+
+"""The longest simple name we allow."""
+MAX_NAME_LENGTH = 100
+
+
+def work_items_from_file(stream):
+    """Yield a series of work items derived from an iterator of lines.
+
+    We expect each line to be a ``str`` with the newline already stripped, as
+    happens automatically with open files, or with ``StringIO`` with universal
+    newline decoding.
+    """
+    items = {}
+    for line in stream:
+        name = hashlib.sha256(line.encode("utf-8")).hexdigest()[-8:]
+        if simple_name_re.match(line.strip()):
+            name += "-" + line.strip()[:MAX_NAME_LENGTH]
```

Comment:
Two things:

* Worth checking for characters that aren't valid in a file name? e.g. if a line contains `/` this will break.
* You're truncate the stripped line to `MAX_NAME_LENGTH` rather than the whole name, so the actual maximum name length is `MAX_NAME_LENGTH + 8`. Is that intentional? If so the variable naming is somewhat confusing.

Reply:
- `test_awkward_lines` already exercises this case, but I don't know what actually happens. Will dig into it.
- Yes it's intentional. I've renamed to `MAX_SIMPLE_NAME_SUFFIX_LENGTH`. Can also change the behaviour to truncate the whole line. I have no preference. Let me know.

Reply:
No need to check for things like `/`. Instead, `simple_name_re` is a whitelist. That means if a line is `some/path`, the path on disk will be `2f4cd884` or something like that.